### DISCUSSION TITLE
mogrify() alternative for query debugging

### DISCUSSION THREAD

**Discussion by sekrause** [psycopg/psycopg#153] [Ideas]
When running a web application that uses SQLAlchemy on top of the old psycopg2, I like to use the following `cursor_factory` in debug mode:

```python
def logging_cursor():
    """
    A custom cursor for the psycopg2 driver which logs all SQL statements in a pretty-printed way.
    """
    from psycopg2.extensions import cursor
    from pygments import highlight
    from pygments.formatter import Formatter
    from pygments.lexers.sql import PostgresLexer
    from pygments_pprint_sql import SqlFilter

    lexer = PostgresLexer()
    lexer.add_filter(SqlFilter())

    class StringFormatter(Formatter):
        def format(self, tokensource, outfile):
            for ttype, value in tokensource:
                outfile.write(value)

    class LoggingCursor(cursor):
        def execute(self, query, vars=None):
            formatted = self.mogrify(query, vars).decode(self.connection.encoding)
            reformatted = highlight(formatted, lexer, StringFormatter())
            indented = textwrap.indent(reformatted.strip(), "    ")
            logging.debug(f"Database query:\n{indented}")
            start = time.perf_counter_ns()
            result = super().execute(query, vars)
            stop = time.perf_counter_ns()
            delta = (stop - start) / 1_000_000
            logging.debug(f"Query execution time: {delta:.3f} ms")
            return result

    return LoggingCursor
```

This `LoggingCursor` uses `mogrify()` to create the query string after arguments binding, reformats the whole query in a pretty-printed way (using [pygments-pprint-sql](https://pypi.org/project/pygments-pprint-sql/)), prints the reformatted query to the debug log and finally executes the original query, measuring its execution time.

Its main use is that I can just copy and paste the whole query (which is mostly autogenerated by SQLAlchemy) into the database GUI of my choice and prefix it with `EXPLAIN ANALYZE` to look the query plan, or maybe to manually modify the query to check why it's not working correctly.

As far as I can tell that wouldn't be possbile anymore with psycopg3 because it uses prepared statements everywhere, so all I can get is the original query with argument placeholders, and a separate parameters list. No more simple copy and paste to re-run the query, I have to merge the statement and its parameters manually.

So what I suggest is that psycopg3 keeps around a function that still does client-side argument binding the old way simply for query debugging purposes.

**Comment by dvarrazzo** [comment]
Hello @sekrause,

The lack of mogrify is caused more by server-side bindings than by prepared statements, but yes, I am aware.

I have thought about this use case: #101 might address it.

Please note that the server wouldn't necessarily process a server-side-bound query exactly the same way of a client-side bound one, so, even if we introduce it, you should either take its results with a pinch of salt (and use extensions such as auto_explain to really see what happened) or execute the client-side-bound query, which might lose some performance but might be perfectly fine for your use cases.

You can get easily a correctly bound query with something like:

```python
>>> import datetime as dt
>>> import psycopg
>>> from psycopg.sql import Literal

>>> query = "select %s, %s, %s"
>>> args = [10, "Hell'o", dt.date(2021,1,1)]

>>> conn = psycopg.connect()
>>> conn.execute(query, args).fetchone()
(10, "Hell'o", datetime.date(2021, 1, 1))

>>> args_str = [Literal(v).as_string(conn) for v in args]
>>> args_str
['10', "'Hell''o'", "'2021-01-01'"]
>>> print(query % tuple(args_str))
select 10, 'Hell''o', '2021-01-01'

>>> conn.execute(query % tuple(args_str)).fetchone()
(10, "Hell'o", '2021-01-01')
```

as you see, in this case, the result is slightly different, as the date doesn't roundtrip, but that's only because it's returned as a constant from the query: if you pass the value to a table or a function argument, where the database has a context, it will perform the right conversion, or you can tweak it using a `::cast`, as you surely know from psql use (this would mostly be needed around overloaded functions such as `generate_series()` which can take different types of arguments).

This is pretty much what #101 would do internally so it's exactly what you'd get out of "mogrify v3".
